// @microsoft/quantum-viz.js/lib/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.regLineStart = exports.argsFontSize = exports.labelFontSize = exports.controlBtnRadius = exports.controlBtnOffset = exports.nestedGroupPadding = exports.groupBoxPadding = exports.classicalRegHeight = exports.registerHeight = exports.labelPadding = exports.gatePadding = exports.gateHeight = exports.minGateWidth = exports.startY = exports.startX = exports.leftPadding = exports.svgNS = void 0;
// SVG Namespace
exports.svgNS = 'http://www.w3.org/2000/svg';
// Display attributes
/** Left padding of SVG. */
exports.leftPadding = 20;
/** x coordinate for first operation on each register. */
exports.startX = 80;
/** y coordinate of first register. */
exports.startY = 40;
/** Minimum width of each gate. */
exports.minGateWidth = 40;
/** Height of each gate. */
exports.gateHeight = 40;
/** Padding on each side of gate. */
exports.gatePadding = 10;
/** Padding on each side of gate label. */
exports.labelPadding = 10;
/** Height between each qubit register. */
exports.registerHeight = exports.gateHeight + exports.gatePadding * 2;
/** Height between classical registers. */
exports.classicalRegHeight = exports.gateHeight;
/** Group box inner padding. */
exports.groupBoxPadding = exports.gatePadding;
/** Padding between nested groups. */
exports.nestedGroupPadding = 2;
/** Additional offset for control button. */
exports.controlBtnOffset = 40;
/** Control button radius. */
exports.controlBtnRadius = 15;
/** Default font size for gate labels. */
exports.labelFontSize = 14;
/** Default font size for gate arguments. */
exports.argsFontSize = 12;
/** Starting x coord for each register wire. */
exports.regLineStart = 40;
