// @microsoft/quantum-viz.js/lib/metadata.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GateType = void 0;
/**
 * Enum for the various gate operations handled.
 */
var GateType;
(function (GateType) {
    /** Measurement gate. */
    GateType[GateType["Measure"] = 0] = "Measure";
    /** CNOT gate. */
    GateType[GateType["Cnot"] = 1] = "Cnot";
    /** SWAP gate. */
    GateType[GateType["Swap"] = 2] = "Swap";
    /** X gate. */
    GateType[GateType["X"] = 3] = "X";
    /** Single/multi qubit unitary gate. */
    GateType[GateType["Unitary"] = 4] = "Unitary";
    /** Single/multi controlled unitary gate. */
    GateType[GateType["ControlledUnitary"] = 5] = "ControlledUnitary";
    /** Nested group of classically-controlled gates. */
    GateType[GateType["ClassicalControlled"] = 6] = "ClassicalControlled";
    /** Group of nested gates */
    GateType[GateType["Group"] = 7] = "Group";
    /** Invalid gate. */
    GateType[GateType["Invalid"] = 8] = "Invalid";
})(GateType = exports.GateType || (exports.GateType = {}));
